lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3551,21 +3600,120 @@ ThreadContext::InvalidatePropertyGuardEntry(const Js::PropertyRecord* propertyRe
 
     entry->uniqueGuards.Clear();
 
-
     // Count no. of invalidations done so far. Exclude if this is all property guards invalidation in which case
     // the unique Guards will be cleared anyway.
     if (!isAllPropertyGuardsInvalidation)
     {
-        this->recyclableData->constructorCacheInvalidationCount += count;
-        if (this->recyclableData->constructorCacheInvalidationCount > (uint)CONFIG_FLAG(ConstructorCacheInvalidationThreshold))
+        this->UpdateConstructorCacheInvalidationCount(count);
+    }
+
+    this->InvalidateEntryPoints(propertyRecord, entry);
+}
+
+void
+ThreadContext::InvalidatePropertyGuards(Js::PropertyId propertyId)
+{
+    const Js::PropertyRecord* propertyRecord = GetPropertyName(propertyId);
+    PropertyGuardDictionary &guards = this->recyclableData->propertyGuards;
+    PropertyGuardEntry* entry = nullptr;
+    if (guards.TryGetValueAndRemove(propertyRecord, &entry))
+    {
+        InvalidatePropertyGuardEntry(propertyRecord, entry, false);
+    }
+}
+
+bool
+ThreadContext::InvalidatePropertyGuardEntryForType(const Js::PropertyRecord* propertyRecord, PropertyGuardEntry* entry, bool isAllPropertyGuardsInvalidation, const Js::Type *type)
+{
+    Assert(entry != nullptr);
+    Assert(type != nullptr);
+
+    // Are we invalidating all entries?  Starts out true; gets set to false if we encounter an entry that we do not invalidate.
+    bool removedAllEntries = true;
+
+    if (entry->sharedGuard != nullptr)
+    {
+        Js::PropertyGuard* guard = entry->sharedGuard;
+        if (guard->GetValue() == reinterpret_cast<intptr_t>(type))
```

Comment:
Can you confirm with @MikeHolman that we don't have a race condition here, since the guards are created and populated by the JIT process? I think we're okay, since we do a final copy on the UI thread before installing the jitted entry point, but I just want to be sure. (Thanks.)